gdnative-core/src/private.rs

```diff
@@ -332,3 +332,28 @@ make_method_table!(struct NativeScriptMethodTable for NativeScript {
 make_method_table!(struct EngineMethodTable for _Engine {
     get_version_info,
 });
+
+static MAIN_THREAD_ID: once_cell::sync::OnceCell<std::thread::ThreadId> =
+    once_cell::sync::OnceCell::new();
+
+#[doc(hidden)]
+pub fn init_main_thread_id() {
+    let c = std::thread::current();
+    MAIN_THREAD_ID
+        .set(c.id())
+        .expect("init_main_thread_id() must be called only once");
+}
+
+#[inline]
+pub(crate) fn assert_main_thread() {
+    debug_assert!(
+        {
+            let c = std::thread::current();
+            let id = MAIN_THREAD_ID
+                .get()
+                .expect("init_main_thread_id() must be called at initialization");
+            c.id() == *id
+        },
+        "Calls from Godot to Rust must be within the main thread"
+    )
+}
```

Comment:
> Calls from Godot to Rust must be within the main thread

Is this absolutely necessary for safety, or something that can be loosened later?

This is a lot more restricting than what we currently allow. Currently, we only limit threaded access to NativeClass objects on a per-object basis by default, and even that can be lifted by specifying a `user_data` wrapper that is `Sync`. To stricten our restrictions to this would constitute a very major change, and would not be mergeable until at least 0.13.

Reply:
I have added the answer to the first post.